pkg/helpers/struct.go

```diff
@@ -208,49 +207,80 @@ func AsSlice(x *messages.ListValue) []interface{} {
 // is possible since they are stored as a float64.
 func NewValue(v interface{}) (*messages.Value, error) {
-	switch v := v.(type) {
-	case nil:
+
+	if v == nil {
 		return NewNullValue(), nil
-	case bool:
-		return NewBoolValue(v), nil
-	case int:
-		return NewNumberValue(float64(v)), nil
-	case int32:
-		return NewNumberValue(float64(v)), nil
-	case int64:
-		return NewNumberValue(float64(v)), nil
-	case uint:
-		return NewNumberValue(float64(v)), nil
-	case uint32:
-		return NewNumberValue(float64(v)), nil
-	case uint64:
-		return NewNumberValue(float64(v)), nil
-	case float32:
-		return NewNumberValue(float64(v)), nil
-	case float64:
-		return NewNumberValue(float64(v)), nil
-	case string:
-		if !utf8.ValidString(v) {
+	}
+
+	switch reflect.TypeOf(v).Kind() {
+	case reflect.Bool:
+		return NewBoolValue(v.(bool)), nil
+	case reflect.Int:
+		return NewNumberValue(float64(v.(int))), nil
+	case reflect.Int32:
+		return NewNumberValue(float64(v.(int32))), nil
+	case reflect.Int64:
+		return NewNumberValue(float64(v.(int64))), nil
+	case reflect.Uint:
+		return NewNumberValue(float64(v.(uint))), nil
+	case reflect.Uint32:
+		return NewNumberValue(float64(v.(uint32))), nil
+	case reflect.Uint64:
+		return NewNumberValue(float64(v.(uint64))), nil
+	case reflect.Float32:
+		return NewNumberValue(float64(v.(float32))), nil
+	case reflect.Float64:
+		return NewNumberValue(v.(float64)), nil
+	case reflect.String:
+		refStr := v.(string)
+		if !utf8.ValidString(refStr) {
 			return nil, protoimpl.X.NewError("invalid UTF-8 in string: %q", v)
 		}
-		return NewStringValue(v), nil
-	case time.Time:
-		return NewTimestampValue(v), nil
-	case []byte:
-		s := base64.StdEncoding.EncodeToString(v)
-		return NewStringValue(s), nil
-	case map[string]interface{}:
-		v2, err := NewStruct(v)
-		if err != nil {
-			return nil, err
+		return NewStringValue(refStr), nil
+	case reflect.Struct:
+		// special case for time values
+		if tVal, ok := v.(time.Time); ok {
+			return NewTimestampValue(tVal), nil
 		}
-		return NewStructValue(v2), nil
-	case []interface{}:
-		v2, err := NewList(v)
+		// should we try to handle struct values somehow?
+		return nil, protoimpl.X.NewError("invalid type: %T", v)
+
+	case reflect.Map:
+		reflected := map[string]interface{}{}
+		if mapV, ok := v.(map[string]interface{}); ok {
+			reflected = mapV
+		} else {
+			mapIter := reflect.ValueOf(v).MapRange()
+			for mapIter.Next() {
+				k := mapIter.Key().String() // This will expect a map of key type string; if we get other key types, there will be some weird values here. Not sure if we want to make that a hard error.
+				mv := mapIter.Value().Interface()
+				reflected[k] = mv
+			}
+		}
+		structVal, err := NewStruct(reflected)
 		if err != nil {
-			return nil, err
+			return nil, protoimpl.X.NewError("could not convert value to struct: %s", err)
 		}
-		return NewListValue(v2), nil
+		return NewStructValue(structVal), nil
+	case reflect.Slice:
```

Comment:
Maybe worth benchmarking this vs just marshalling to JSON and then calling https://pkg.go.dev/google.golang.org/protobuf/types/known/structpb#ListValue.UnmarshalJSON. The JSON package might special case some of the reflection better than we can.

Reply:
So, it looks like for most things, marshalling to JSON and then unarshalling via the `protojson.Unmarshal` is slower:
```
# code in this PR
go test -bench=BenchmarkCustomUnmarshal -count 5 -benchmem
goos: linux
goarch: amd64
pkg: github.com/elastic/elastic-agent-shipper-client/pkg/helpers
cpu: Intel(R) Xeon(R) CPU E5-2630 0 @ 2.30GHz
BenchmarkCustomUnmarshal-12       359370              4787 ns/op            1000 B/op         20 allocs/op
BenchmarkCustomUnmarshal-12       258730              4829 ns/op            1000 B/op         20 allocs/op
BenchmarkCustomUnmarshal-12       224470              4720 ns/op            1000 B/op         20 allocs/op
BenchmarkCustomUnmarshal-12       214449              4678 ns/op            1000 B/op         20 allocs/op
BenchmarkCustomUnmarshal-12       218415              4641 ns/op            1000 B/op         20 allocs/op
PASS
ok      github.com/elastic/elastic-agent-shipper-client/pkg/helpers     7.318s

# marshalling to JSON then unmarshalling via the structpb UnmarshalJSON()
 go test -bench=BenchmarkUpstream -count 5 -benchmem                
goos: linux
goarch: amd64
pkg: github.com/elastic/elastic-agent-shipper-client/pkg/helpers
cpu: Intel(R) Xeon(R) CPU E5-2630 0 @ 2.30GHz
BenchmarkUpstream-12               61833             26842 ns/op            1848 B/op         48 allocs/op
BenchmarkUpstream-12               40722             27175 ns/op            1849 B/op         48 allocs/op
BenchmarkUpstream-12               52257             26778 ns/op            1848 B/op         48 allocs/op
BenchmarkUpstream-12               40884             26061 ns/op            1848 B/op         48 allocs/op
BenchmarkUpstream-12               40881             25259 ns/op            1848 B/op         48 allocs/op
PASS
ok      github.com/elastic/elastic-agent-shipper-client/pkg/helpers     7.621s
```

Reply:
Yeah that is a big difference. So the conclusion is to keep the custom code you added, but we should fall back to json.Marshal instead of reporting an error perhaps with a debug log when that happens. I would rather be slow than broken.

Also please commit the benchmarks they used, they are really valuable for resolving questions in here.

Reply:
Alright, I added a fallback for the struct handling at least. Not sure how aggressive we want to be with falling back to some other unmarshalling scheme; as it stands right now, one of the only types we'll fall back to is maps that have a key type other than string, and I feel like the type shenanigans involved in turning `map[int]interface{}` into `map[string]interface{}` (and so on) probably do warrant an error.